src/sundials/sundials_errors.c

```diff
@@ -0,0 +1,91 @@
+/* -----------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2022, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------*/
+
+#include <stdlib.h>
+#include <string.h>
+#include <sundials/sundials_core.h>
+#include <sundials/impl/sundials_errors_impl.h>
+
+static inline char* combineFileAndLine(int line, const char* file)
+{
+  size_t total_str_len = strlen(file) + 6; /* TODO(CJB): need to figure out width of line */
+  char* file_and_line  = malloc(total_str_len * sizeof(char));
+  snprintf(file_and_line, total_str_len, "%s:%d", file, line);
+  return file_and_line;
+}
+
+SUNErrHandler SUNErrHandler_Create(SUNErrHandlerFn eh_fn, void* eh_data)
+{
+  SUNErrHandler eh = NULL;
+  eh               = (SUNErrHandler)malloc(sizeof(struct SUNErrHandler_));
+  eh->previous     = NULL;
+  eh->call         = eh_fn;
+  eh->data         = eh_data;
+  return eh;
+}
+
+void SUNErrHandler_Destroy(SUNErrHandler eh)
+{
+  while (eh != NULL) {
+    SUNErrHandler next_eh = eh->previous;
+    free(eh);
+    eh = next_eh;
+  }
+}
+
+const char* SUNGetErrMsg(SUNErrCode code, SUNContext sunctx)
+{
+#define SUN_EXPAND_TO_CASES(name, description) \
+  case name: return description; break;
+
+  switch (code) {
+    SUN_ERR_CODE_LIST(SUN_EXPAND_TO_CASES)
+  default: return "unknown error";
```

Comment:
Seems like this could lead to a dangling pointer. This may need an out argument instead

Reply:
Not sure I am following how this would cause a dangling pointer. `SUN_ERR_CODE_LIST` should expand to a bunch of statements like:
```
case(ERR_CODE): return "some string";
```
Returning a string literal as a `const char*` is perfectly valid since string literals have static duration.

Reply:
Ahh that's right